packages/api/internal/orchestrator/nodes/node.go

```diff
@@ -0,0 +1,185 @@
+	}
+
+	nodeStatus := api.NodeStatusUnhealthy
+	nodeInfo, err := client.Info.ServiceInfo(ctx, &emptypb.Empty{})
+	if err != nil {
+		_ = client.Close()
+		return nil, fmt.Errorf("failed to get node service info: %w", err)
+	}
+
+	nodeStatus, ok := OrchestratorToApiNodeStateMapper[nodeInfo.ServiceStatus]
+	if !ok {
+		zap.L().Error("Unknown service info status", zap.Any("status", nodeInfo.ServiceStatus), logger.WithNodeID(nodeInfo.NodeId))
+		nodeStatus = api.NodeStatusUnhealthy
+	}
+
+	buildCache := ttlcache.New[string, interface{}]()
+	go buildCache.Start()
+
+	nodeMetadata := NodeMetadata{
+		serviceInstanceID: nodeInfo.ServiceId,
+		Commit:            nodeInfo.ServiceCommit,
+		Version:           nodeInfo.ServiceVersion,
+	}
+
+	clusterID := uuid.Nil
+	n := &Node{
+		NomadNodeShortID: discoveredNode.NomadNodeShortID,
+		ClusterID:        clusterID,
+		ID:               nodeInfo.NodeId,
+		IPAddress:        discoveredNode.IPAddress,
+
+		client: client,
+		status: nodeStatus,
+		meta:   nodeMetadata,
+
+		buildCache: buildCache,
+		PlacementMetrics: PlacementMetrics{
+			sandboxesInProgress: smap.New[*sbxInProgress](),
+			createSuccess:       atomic.Uint64{},
+			createFails:         atomic.Uint64{},
+		},
+	}
+	n.UpdateMetricsFromServiceInfoResponse(nodeInfo)
+
+	return n, nil
+}
+
+func NewClusterNode(
+	ctx context.Context,
+	client *grpclient.GRPCClient,
+	clusterID uuid.UUID,
+	i *edge.ClusterInstance,
+) (*Node, error) {
+	nodeStatus, ok := OrchestratorToApiNodeStateMapper[i.GetStatus()]
+	if !ok {
+		zap.L().Error("Unknown service info status", zap.Any("status", i.GetStatus()), logger.WithNodeID(i.NodeID))
+		nodeStatus = api.NodeStatusUnhealthy
+	}
+
+	buildCache := ttlcache.New[string, interface{}]()
+	go buildCache.Start()
+
+	nodeMetadata := NodeMetadata{
+		serviceInstanceID: i.ServiceInstanceID,
+		Commit:            i.ServiceVersionCommit,
+		Version:           i.ServiceVersion,
+	}
+
+	n := &Node{
+		NomadNodeShortID: UnknownNomadNodeShortID,
+		ClusterID:        clusterID,
+		ID:               i.NodeID,
+		// We can't connect directly to the node in the cluster
+		IPAddress: "",
+
+		client: client,
+		status: nodeStatus,
+		meta:   nodeMetadata,
+
+		buildCache: buildCache,
+		PlacementMetrics: PlacementMetrics{
+			sandboxesInProgress: smap.New[*sbxInProgress](),
+			createSuccess:       atomic.Uint64{},
+			createFails:         atomic.Uint64{},
+		},
+	}
+
+	nodeInfo, err := client.Info.ServiceInfo(ctx, &emptypb.Empty{})
+	if err != nil {
+		zap.L().Error("Failed to get node service info", zap.Error(err), logger.WithNodeID(n.ID))
+		return n, nil
+	}
+
+	n.UpdateMetricsFromServiceInfoResponse(nodeInfo)
+
+	return n, nil
+}
+
+func (n *Node) Close() error {
+	n.buildCache.Stop()
+
+	return n.client.Close()
+}
+
+func (n *Node) CloseWithClient() error {
+	err := n.client.Close()
+	if err != nil {
+		zap.L().Error("Error closing connection to node", zap.Error(err), logger.WithNodeID(n.ID))
+	}
+
+	return n.Close()
+}
```

Comment:
What `Close()` also canceling the client connection? It should not be done for clusters as the connection is shared for the whole cluster.

A close call is missed during cluster node removal.